pageserver/src/tenant/timeline.rs

```diff
@@ -3890,9 +3872,28 @@ impl Timeline {
         // image layers  <100000000..100000099> and <200000000..200000199> are not completely covering it.
         let mut start = Key::MIN;
 
+        let check_for_image_layers = {
+            let last_checks_at = self.last_image_layer_creation_check_at.load();
+            let distance = lsn
+                .checked_sub(last_checks_at)
+                .expect("Attempt to compact with LSN going backwards");
+            let min_distance = self.get_image_layer_creation_check_threshold() as u64
+                * self.get_checkpoint_distance();
+
+            // Skip the expensive delta layer counting if this timeline has not ingested sufficient
+            // WAL since the last check.
+            distance.0 >= min_distance
+        };
+
+        if check_for_image_layers {
+            self.last_image_layer_creation_check_at.store(lsn);
+        }
+
         for partition in partitioning.parts.iter() {
             let img_range = start..partition.ranges.last().unwrap().end;
-            if !force && !self.time_for_new_image_layer(partition, lsn).await {
+            if !force
+                && (!check_for_image_layers || !self.time_for_new_image_layer(partition, lsn).await)
+            {
```

Comment:
Many ! , and short-circuited evaluation make my head hurt.

Feel free to dismiss

```suggestion
            // time_for_new_image_layer alone is somewhat expensive, even if we don't generate image layers
            let doit = if force {
                true
            } else {
                if check_for_image_layers {
                    self.time_for_new_image_layer(partition, lsn).await
                } else {
                    false
                }
            };
            if !doit {
```

Reply:
Done in https://github.com/neondatabase/neon/pull/7420/commits/aa2ea712e0ede016641e2d9b8d754c8beeab3e68